pkg/ccl/changefeedccl/schemafeed/schema_feed.go

```diff
@@ -481,27 +470,43 @@ func (tf *schemaFeed) pauseOrResumePolling(
 		return atOrBefore, nil
 	}
 
-	if tf.mu.allTableVersions1 == nil {
-		tf.mu.allTableVersions1 = make(map[descpb.ID]descpb.DescriptorVersion)
-		tf.mu.allTableVersions2 = make(map[descpb.ID]descpb.DescriptorVersion)
-	}
-
 	// Always start with a stance to resume polling until we've proved otherwise.
 	tf.mu.pollingPaused = false
-	if ok, err := areAllLeasedTablesSchemaLockedAt(tf.mu.highWater, tf.mu.allTableVersions1); err != nil || !ok {
-		return atOrBefore, err
-	}
-	if ok, err := areAllLeasedTablesSchemaLockedAt(atOrBefore, tf.mu.allTableVersions2); err != nil || !ok {
-		return atOrBefore, err
-	}
-	if len(tf.mu.allTableVersions1) != len(tf.mu.allTableVersions2) {
-		return atOrBefore, nil
-	}
-	for id, version := range tf.mu.allTableVersions1 {
-		if version != tf.mu.allTableVersions2[id] {
-			return atOrBefore, nil
+
+	// Determine if table descriptor versions at highWater and at atOrBefore are
+	// both locked, and have the same versions.
+	allSame, err := func() (bool, error) {
+		for _, id := range tf.tableIDs {
+			// Use previously verified state.
+			highWaterVersion := tf.mu.lockedVersions[id]
+			if tf.mu.versionsLockedAt.Less(tf.mu.highWater) {
+				// Need to re-load locked versions at highWater.
+				v, isLocked, err := tableDescriptorStateAt(id, tf.mu.highWater)
+				if err != nil || !isLocked {
+					return false, err
+				}
+				tf.mu.lockedVersions[id] = v
+				highWaterVersion = v
```

Comment:
you are iterating over all watched tables here. You used `t.mu.versionsLockedAt` in the check condition to determine whether you should "re-load locked versions at highWater", but you update `tf.mu.versionsLockedAt` too early inside the check. Wouldn't this be buggy bc if there are two watched tables, and we need to re-load both tables at `tf.mu.highWater`, the IF condition for the second table will not be true because you prematurely updated `tf.mu.versionsLockedAt` when loading the first watched table.

Reply:
Most excellent point! Now, I remember why I had this update outside for loop  :) 
Good catch.

Reply:
now that `tf.mu` is locked when we are here (and thus `tf.mu.highwater` cannot change), and `tf.mu.versionsLockedAt` also won't change when we are iterating over all watched tables, then we can move this check (`if tf.mu.versionsLockedAt.Less(tf.mu.highWater)`) outside of the for-each-table-loop. In other words, we just need to check this condition (`if tf.mu.versionsLockedAt.Less(tf.mu.highWater)`) once, and if true, we re-load all tables.